src/arch/helperrvv.h

```diff
@@ -46,52 +46,56 @@
 #define ENABLE_DP
 
 #if CONFIG != 2
+#if defined(ENABLE_RVVM1NOFMA) || defined(ENABLE_RVVM2NOFMA)
+#error "RVV NOFMA only supported for CONFIG=2"
+#else
 #define ENABLE_FMA_SP
 #define ENABLE_FMA_DP
 #endif
+#endif
 
-static INLINE int vavailability_i(int name) { return -1; }
-
-////////////////////////////////////////////////////////////////////////////////
-// RISC-V Vector Types
-////////////////////////////////////////////////////////////////////////////////
+#if __riscv_v_intrinsic <= 12000
+#define __riscv_vcreate_v_f32m1_f32m2(x, y) __riscv_vset(__riscv_vlmul_ext_v_f32m1_f32m2(x), 1, y)
+#define __riscv_vcreate_v_f32m2_f32m4(x, y) __riscv_vset(__riscv_vlmul_ext_v_f32m2_f32m4(x), 1, y)
+#define __riscv_vcreate_v_f32m4_f32m8(x, y) __riscv_vset(__riscv_vlmul_ext_v_f32m4_f32m8(x), 1, y)
+#define __riscv_vcreate_v_f64m1_f64m2(x, y) __riscv_vset(__riscv_vlmul_ext_v_f64m1_f64m2(x), 1, y)
+#define __riscv_vcreate_v_f64m2_f64m4(x, y) __riscv_vset(__riscv_vlmul_ext_v_f64m2_f64m4(x), 1, y)
+#define __riscv_vcreate_v_f64m4_f64m8(x, y) __riscv_vset(__riscv_vlmul_ext_v_f64m4_f64m8(x), 1, y)
+#define __riscv_vcreate_v_i32m1_i32m2(x, y) __riscv_vset(__riscv_vlmul_ext_v_i32m1_i32m2(x), 1, y)
+#define __riscv_vcreate_v_i32m2_i32m4(x, y) __riscv_vset(__riscv_vlmul_ext_v_i32m2_i32m4(x), 1, y)
+#define __riscv_vcreate_v_i32m4_i32m8(x, y) __riscv_vset(__riscv_vlmul_ext_v_i32m4_i32m8(x), 1, y)
+#define __riscv_vcreate_v_u64m1_u64m2(x, y) __riscv_vset(__riscv_vlmul_ext_v_u64m1_u64m2(x), 1, y)
+#define __riscv_vcreate_v_u64m2_u64m4(x, y) __riscv_vset(__riscv_vlmul_ext_v_u64m2_u64m4(x), 1, y)
 
-// About the RISC-V Vector type translations:
-//
-// Because the single- and double-precision versions of the RVV port have
-// conflicting definitions of the vmask and vopmask types, they can only
-// be defined for at most one precision level in a single translation unit.
-// Any functions that use vmask or vopmask types are thus enabled only by the
-// corresponding ENABLE_RVV_SP or ENABLE_RVV_DP macro guards.
-#if defined(ENABLE_RVV_SP) && defined(ENABLE_RVV_DP)
-#error Cannot simultaneously define ENABLE_RVV_SP and ENABLE_RVV_DP
+#define SLEEF_RVV_UNINITIALIZED_REG(T) __extension__({          \
```

Comment:
The extension used here is supported by GCC and Clang, but perhaps it may be in issue for other compilers.

If you recall that `(a, b, c)` evaluates a, b, and c, and returns c, this extension with `{}` inside of the `()` and allows more complex logic, and the last whole statement is the result (the pragma doesn't count).  The `__extension__` keyword just clears this with the linter saying that I have not forgotten that it doesn't conform to standard C (also supported in both GCC and Clang).

I forgot that I went outside of standard C, to be honest.  It was hammered out in anger when I discovered the version incompatibility.

Before I go and write something to that effect in comments, are we OK with use of a GCC/Clang extension which might interfere with other compilers?  They can avoid this problem just as soon as they adopt the 1.0 version of the intrinsics specification and this workaround won't be used.

Reply:
Looking back at how it's used; I made the macro expecting to have to use it many more times than I eventually did.

Since there are only two of them, it's simpler to just use a couple of inline functions.

Sadly the pragmas are stripped by the preprocessor for inline headers (both #pragma and _Pragma are stripped), so when it comes to generating inline headers the warnings will come back.  When the toolchain is updated this nit also goes away.

Reply:
> Before I go and write something to that effect in comments, are we OK with use of a GCC/Clang extension which might interfere with other compilers?

I think this is an acceptable limitation since the support for other compilers retrograded to experimental. Is it still the case with the inline functions?

> Sadly the pragmas are stripped by the preprocessor for inline headers (both #pragma and _Pragma are stripped), so when it comes to generating inline headers the warnings will come back. When the toolchain is updated this nit also goes away.

Does it not help to add `//@#pragma ...`? I suppose we can deal with a few warnings until the toolchain is updated.

If you find a way to avoid that altogether in the future that's obviously preferable, but for now I'm happy with the most recent change.

Reply:
I just tried it some `//@` tricks, but it looks like the preprocessor eats up the pragmas as soon as they appear, leaving no opportunity to squeeze them through to the next layer.

There might be something more esoteric with token pasting and/or conditional code that recognises that the file is being preprocessed, but if I ever figure that out I think it would be better to do it in a separate tidy-up patch.

So as things stand this is the best I know how to do.

Having replaced the macros with inline functions they should now be portable to any compiler again.  And the rules for pragmas is that if a compiler doesn't know it it just ignores it (hence GCC putting their own name at the front to isolate everything under that to their own namespace so there are no collisions).

I think this should be good to go, now?

Reply:
Ok, looks good to me